packages/api/engine/opensea.go

```diff
@@ -0,0 +1,155 @@
+package engine
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net"
+	"net/http"
+	"sync"
+	"time"
+)
+
+// Opensea is an opensea client
+type Opensea struct {
+	sync.Mutex
+	API      string
+	APIKey   string
+	contract string
+	ticker   *time.Ticker
+}
+
+type errorResponse struct {
+	Success bool `json:"success"`
+}
+
+func (e errorResponse) Error() string {
+	return "Not success"
+}
+
+// OpenseaConfig config
+type OpenseaConfig struct {
+	URL      string
+	Interval time.Duration
+	APIKey   string
+	contract string
+}
+
+// NewOpensea creates an opensea API client on maassetet
+func NewOpensea(config OpenseaConfig) *Opensea {
+	o := &Opensea{
+		API:      config.URL,
+		contract: config.contract,
+		ticker:   time.NewTicker(config.Interval),
+	}
+	return o
+}
+
+// Sync implemented for Opensa
+func (o *Opensea) Sync(ctx context.Context) {
+	defer o.ticker.Stop()
+	fmt.Printf("Opensea API is %s", o.API)
+	for {
+		select {
+		case <-o.ticker.C:
+			fmt.Printf("in for API is %s", o.API)
+			o.Lock()
+
+			ret, _ := o.GetAssetCollection(o.contract)
+			fmt.Println(len(ret.Assets))
+
+			for _, asset := range ret.Assets {
+				// Implement save to current Asset and Listing model.
+				fmt.Printf("tokenid below")
+				fmt.Println(asset.TokenID)
+				fmt.Printf("Getting current price")
+				fmt.Println(asset.SellOrders)
+				if asset.LastSale != nil {
+					fmt.Printf("Get last sale price from here")
+					fmt.Println(asset.LastSale.TotalPrice)
+				}
+
+			}
+			o.Unlock()
+		case <-ctx.Done():
+			return
+		}
+	}
+}
+
+// GetAssetCollection for Opensa
+func (o Opensea) GetAssetCollection(assetContractAddress string) (*Assets, error) {
+	ctx := context.TODO()
+	return o.GetAssetCollectionWithContext(ctx, assetContractAddress)
+}
+
+// GetAssetCollectionWithContext for Opensa
+func (o Opensea) GetAssetCollectionWithContext(ctx context.Context, assetContractAddress string) (*Assets, error) {
+	path := fmt.Sprintf("/api/v1/assets?asset_contract_address=%s&order_direction=asc&limit=%v", assetContractAddress, 10)
```

Comment:
```suggestion
	path := fmt.Sprintf("/api/v1/assets?asset_contract_address=%s&order_direction=asc&limit=%d", assetContractAddress, 10)
```

Reply:
Modified this but going to have to paginate 50 records at a time from Opensea.

Reply:
Sounds good. I also wonder if we might consider refreshing active listings more frequently than the entire collection. So if it is active, refresh every 5mins rather than 1hr